ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/ParseReportingUrl.java

```diff
@@ -87,40 +99,104 @@ public Result<PCollection<PubsubMessage>, PubsubMessage> expand(
           }
 
           Map<String, String> attributes = new HashMap<>(message.getAttributeMap());
+          String namespace = Objects
+              .requireNonNull(message.getAttribute(Attribute.DOCUMENT_NAMESPACE));
+          String docType = Objects.requireNonNull(message.getAttribute(Attribute.DOCUMENT_TYPE));
+
+          SponsoredInteraction.Builder interactionBuilder = SponsoredInteraction.builder();
+
+          interactionBuilder.setOriginalNamespace(namespace);
+          interactionBuilder.setOriginalDocType(docType);
+
+          // set fields based on namespace/doctype combos
+          if (NS_DESKTOP.equals(namespace)) {
+            interactionBuilder.setFormFactor(SponsoredInteraction.FORM_DESKTOP);
+            // potential docTypes here are
+            // `topsites-impression`, `topsites-click`
+            // `quicksuggest-impression`, `quicksuggest-click`
+            if (DT_TOPSITES_IMPRESSION.equals(docType)
+                || DT_QUICKSUGGEST_IMPRESSION.equals(docType)) {
+              interactionBuilder.setInteractionType(SponsoredInteraction.INTERACTION_IMPRESSION);
+            } else if (DT_TOPSITES_CLICK.equals(docType) || DT_QUICKSUGGEST_CLICK.equals(docType)) {
+              interactionBuilder.setInteractionType(SponsoredInteraction.INTERACTION_CLICK);
+            } else {
+              throw new InvalidAttributeException("Received unexpected docType: " + docType,
+                  docType);
+            }
+          } else {
+            interactionBuilder.setFormFactor(SponsoredInteraction.FORM_PHONE);
+            // enforce that the only mobile docType is `topsites-impression`
+            if (!DT_TOPSITES_IMPRESSION.equals(docType)) {
+              throw new InvalidAttributeException("Unexpected docType for mobile ping: " + docType,
+                  docType);
+            }
+            ArrayNode events = payload.withArray("events");
+            if (events.size() != 1) {
+              throw new RuntimeException("expect exactly 1 event in ping.");
```

Comment:
`InvalidAttributeException` might be okay here. We should make sure we use something that is caught and sent to error output rather than letting this bubble up and take down the pipeline.